### PR TITLE
LeastRequestedPriority/MostRequestedPriority/BalancedResourceAllocation as Score plugins

### PR DESCRIPTION
**Pull request by zouyee**
/assign @ahg-g

**What type of PR is this?**
/kind feature
/priority important-soon

**What this PR does / why we need it**:
 LeastRequestedPriority/MostRequestedPriority/BalancedResourceAllocation as Score plugins

**Which issue(s) this PR fixes**:

Part of #84059
**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: pkg/scheduler/framework/plugins/nodeutilization/balanced_test.go [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	featuregatetesting "k8s.io/component-base/featuregate/testing"
+	"k8s.io/kubernetes/pkg/features"
+	framework "k8s.io/kubernetes/pkg/scheduler/framework/v1alpha1"
+	schedulernodeinfo "k8s.io/kubernetes/pkg/scheduler/nodeinfo"
+)
+
+func TestBalancedResourceAllocation(t *testing.T) {
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/nodeutilization/balanced_test.go]: update test name

File: pkg/scheduler/framework/plugins/nodeutilization/balanced_test.go [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+			/*
+				Node1 scores (remaining resources) on 0-10 scale
+				CPU Fraction: 0 / 4000 = 0%
+				Memory Fraction: 0 / 10000 = 0%
+				Node1 Score: 10 - (0-0)*100 = 100
+				Node2 scores (remaining resources) on 0-10 scale
+				CPU Fraction: 0 / 4000 = 0 %
+				Memory Fraction: 0 / 10000 = 0%
+				Node2 Score: 10 - (0-0)*100 = 100
+			*/
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/nodeutilization/balanced_test.go]: change comment format to ```//```, here and all cases below please.

File: pkg/scheduler/framework/plugins/nodeutilization/high.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/kubernetes/pkg/scheduler/algorithm/priorities"
+	"k8s.io/kubernetes/pkg/scheduler/framework/plugins/migration"
+	framework "k8s.io/kubernetes/pkg/scheduler/framework/v1alpha1"
+)
+
+// HighNodeUtilization is a plugin that favors nodes with most requested resources.
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/nodeutilization/high.go]: "HighNodeUtilization is a plugin that favors nodes with high utilization based on requested resources."

File: pkg/scheduler/framework/plugins/nodeutilization/high_test.go [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+	}
+	tests := []struct {
+		pod          *v1.Pod
+		pods         []*v1.Pod
+		nodes        []*v1.Node
+		expectedList framework.NodeScoreList
+		name         string
+	}{
+		{
+			/*
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/nodeutilization/high_test.go]: ditto comment format.

File: pkg/scheduler/framework/plugins/nodeutilization/low_test.go [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+	}
+	tests := []struct {
+		pod          *v1.Pod
+		pods         []*v1.Pod
+		nodes        []*v1.Node
+		expectedList framework.NodeScoreList
+		name         string
+	}{
+		{
+			/*
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/nodeutilization/low_test.go]: ditto comment format.

File: pkg/scheduler/framework/plugins/nodeutilization/test_util.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+			},
+			Allocatable: v1.ResourceList{
+				v1.ResourceCPU:    *resource.NewMilliQuantity(milliCPU, resource.DecimalSI),
+				v1.ResourceMemory: *resource.NewQuantity(memory, resource.BinarySI),
+			},
+		},
+	}
+}
+
+// getExistingVolumeCountForNode gets the current number of volumes on node.
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/nodeutilization/test_util.go]: this is used in one test, we should probably put it in that test only.

**Review by ahg-g — COMMENTED** [review]
> Signed-off-by: Zou Nengren [zouyee1989@gmail.com](mailto:zouyee1989@gmail.com)
> 
> /assign @ahg-g
> 
> **What type of PR is this?**
> /kind feature
> /priority important-soon
> 
> **What this PR does / why we need it**:
> LeastRequestedPriority/MostRequestedPriority/BalancedResourceAllocation as Score plugins
> 
> **Which issue(s) this PR fixes**:
> 
> Part of #84059
> **Special notes for your reviewer**:
> 
> **Does this PR introduce a user-facing change?**:
> 
> ```
>  LeastRequestedPriority/MostRequestedPriority/BalancedResourceAllocation as Score plugins
> ```
> 
> **Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
> 
> ```
> 
> ```

This is not a user facing change. Please just put NONE in the release notes.

**Comment by ahg-g** [comment]
I am honestly not quite sure about the naming, "utilization" may not be accurate. I think "allocation" is the more accurate term here because the scheduler know only how much of a node's resources was allocated, but not the actual utilization. Let me try to come up with a better naming.

**Comment by ahg-g** [comment]
Here is one idea:

we place them all under ```NodeResources```:

1) MostRequestedPriority would be under ```NodeResources/most_allocated.go```, and the plugin name would be ```NodeResourcesMostAllocated```

2) LeastRequestedPriority would be under ```noderesources/least_allocated.go```, and the plugin name would be ```NodeResourcesLeastAllocated```

3) BalancedResourceAllocation would be under ```noderesources/least_allocated.go```, and the plugin name would be ```NodeResourcesBalancedAllocation```

4) in a followup PR, we move ```noderesources/node_resources.go``` to ```noderesources/fit.go```  and the plugin name would be ```NodeResourcesFit```

File: pkg/scheduler/framework/plugins/noderesources/most_allocated.go [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+// NodeResourcesMostAllocated is a score plugin that favors nodes with high allocation based on requested resources.
+type NodeResourcesMostAllocated struct {
+	handle framework.FrameworkHandle
+}
+
+var _ = framework.ScorePlugin(&NodeResourcesMostAllocated{})
+
+// NodeResourcesMostAllocatedName is the name of the plugin used in the plugin registry and configurations.
+const NodeResourcesMostAllocatedName = "NodeResourcesMostAllocated"
```
Comment (draveness) [inline pkg/scheduler/framework/plugins/noderesources/most_allocated.go]: We could cut the noderesources from the constant and struct, `noderesoures.MostAllocatedName` could be better compare to `noderesources.NdoeResourcesMostAllocatedName`

File: pkg/scheduler/framework/plugins/noderesources/most_allocated.go [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+	s, err := priorities.MostRequestedPriorityMap(pod, meta, nodeInfo)
+	return s.Score, migration.ErrorToFrameworkStatus(err)
+}
+
+// ScoreExtensions of the Score plugin.
+func (nrma *NodeResourcesMostAllocated) ScoreExtensions() framework.ScoreExtensions {
+	return nil
+}
+
+// NewHighNodeUtilization initializes a new plugin and returns it.
```
Comment (draveness) [inline pkg/scheduler/framework/plugins/noderesources/most_allocated.go]: `noderesources.NewMostAllocated` could be a better name

File: pkg/scheduler/framework/plugins/noderesources/least_allocated.go [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+// NodeResourcesLeastAllocated is a score plugin that favors nodes with fewer allocation requested resources based on requested resources.
+type NodeResourcesLeastAllocated struct {
+	handle framework.FrameworkHandle
+}
+
+var _ = framework.ScorePlugin(&NodeResourcesLeastAllocated{})
+
+// NodeResourcesLeastAllocatedName is the name of the plugin used in the plugin registry and configurations.
+const NodeResourcesLeastAllocatedName = "NodeResourcesLeastAllocated"
```
Comment (draveness) [inline pkg/scheduler/framework/plugins/noderesources/least_allocated.go]: ditto

File: pkg/scheduler/framework/plugins/noderesources/balance_allocated.go [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+// NodeResourcesBalancedAllocation is a score plugin that calculates the difference between the cpu and memory fraction
+// of capacity, and prioritizes the host based on how close the two metrics are to each other.
+type NodeResourcesBalancedAllocation struct {
+	handle framework.FrameworkHandle
+}
+
+var _ = framework.ScorePlugin(&NodeResourcesBalancedAllocation{})
+
+// NodeResourcesBalancedAllocationName is the name of the plugin used in the plugin registry and configurations.
+const NodeResourcesBalancedAllocationName = "NodeResourcesBalancedAllocation"
```
Comment (draveness) [inline pkg/scheduler/framework/plugins/noderesources/balance_allocated.go]: ditto

**Comment by draveness** [comment]
The golint also complains about the naming here.

```go
pkg/scheduler/framework/plugins/noderesources/least_allocated.go:28:6: type name will be used as noderesources.NodeResourcesLeastAllocated by other packages, and that stutters; consider calling this LeastAllocated
pkg/scheduler/framework/plugins/noderesources/most_allocated.go:28:6: type name will be used as noderesources.NodeResourcesMostAllocated by other packages, and that stutters; consider calling this MostAllocated
```

File: pkg/scheduler/framework/plugins/noderesources/balance_allocated.go [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+	handle framework.FrameworkHandle
+}
+
+var _ = framework.ScorePlugin(&NodeResourcesBalancedAllocation{})
+
+// BalancedAllocationName is the name of the plugin used in the plugin registry and configurations.
+const BalancedAllocationName = "NodeResourcesBalancedAllocation"
+
+// Name returns name of the plugin. It is used in logs, etc.
+func (bnu *NodeResourcesBalancedAllocation) Name() string {
```
Comment (draveness) [inline pkg/scheduler/framework/plugins/noderesources/balance_allocated.go]: BalancedAllocation

File: pkg/scheduler/framework/plugins/noderesources/least_allocated.go [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/kubernetes/pkg/scheduler/algorithm/priorities"
+	"k8s.io/kubernetes/pkg/scheduler/framework/plugins/migration"
+	framework "k8s.io/kubernetes/pkg/scheduler/framework/v1alpha1"
+)
+
+// NodeResourcesLeastAllocated is a score plugin that favors nodes with fewer allocation requested resources based on requested resources.
+type NodeResourcesLeastAllocated struct {
```
Comment (draveness) [inline pkg/scheduler/framework/plugins/noderesources/least_allocated.go]: LeastAllocated

File: pkg/scheduler/framework/plugins/noderesources/most_allocated.go [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/kubernetes/pkg/scheduler/algorithm/priorities"
+	"k8s.io/kubernetes/pkg/scheduler/framework/plugins/migration"
+	framework "k8s.io/kubernetes/pkg/scheduler/framework/v1alpha1"
+)
+
+// NodeResourcesMostAllocated is a score plugin that favors nodes with high allocation based on requested resources.
+type NodeResourcesMostAllocated struct {
```
Comment (draveness) [inline pkg/scheduler/framework/plugins/noderesources/most_allocated.go]: MostAllocated

**Review by draveness — COMMENTED** [review]
please take a look at the failing CI

File: pkg/scheduler/framework/plugins/noderesources/least_allocated.go [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+	handle framework.FrameworkHandle
+}
+
+var _ = framework.ScorePlugin(&NodeResourcesLeastAllocated{})
+
+// LeastAllocatedName is the name of the plugin used in the plugin registry and configurations.
+const LeastAllocatedName = "NodeResourcesLeastAllocated"
+
+// Name returns name of the plugin. It is used in logs, etc.
+func (nrla *NodeResourcesLeastAllocated) Name() string {
```
Comment (draveness) [inline pkg/scheduler/framework/plugins/noderesources/least_allocated.go]:
```suggestion
func (pl *NodeResourcesLeastAllocated) Name() string {
```

Same as other plugins

**Review by draveness — CHANGES_REQUESTED** [review]

**Comment by zouyee** [comment]
@ahg-g PTAL

**Comment by ahg-g** [comment]
looks good, I will approve once https://github.com/kubernetes/kubernetes/pull/84066 is merged.

**Comment by draveness** [comment]
@zouyee Please rebase

File: pkg/scheduler/framework/plugins/noderesources/balance_allocated.go [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+	return BalancedAllocationName
+}
+
+// Score invoked at the score extension point.
+func (ba *BalancedAllocation) Score(ctx context.Context, state *framework.CycleState, pod *v1.Pod, nodeName string) (int64, *framework.Status) {
+	nodeInfo, exist := ba.handle.NodeInfoSnapshot().NodeInfoMap[nodeName]
+	if !exist {
+		return 0, framework.NewStatus(framework.Error, fmt.Sprintf("node %q does not exist in NodeInfoSnapshot", nodeName))
+	}
+	meta := migration.PriorityMetadata(state)
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/noderesources/balance_allocated.go]: sorry I missed this, BalancedResourceAllocationMap does not use meta, so this lookup is redundant. please remove it and pass nil instead, and add a comment ```"BalancedResourceAllocationMap does not use priority metadata, hence we pass nil here"```

File: pkg/scheduler/framework/plugins/noderesources/balance_allocated.go [hunk 16]
```diff
@@ -0,0 +1,66 @@
+/*
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/noderesources/balance_allocated.go]: file name should be ```balanced_allocation```

File: pkg/scheduler/framework/plugins/noderesources/least_allocated.go [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+	return LeastAllocatedName
+}
+
+// Score invoked at the score extension point.
+func (la *LeastAllocated) Score(ctx context.Context, state *framework.CycleState, pod *v1.Pod, nodeName string) (int64, *framework.Status) {
+	nodeInfo, exist := la.handle.NodeInfoSnapshot().NodeInfoMap[nodeName]
+	if !exist {
+		return 0, framework.NewStatus(framework.Error, fmt.Sprintf("node %q does not exist in NodeInfoSnapshot", nodeName))
+	}
+	meta := migration.PriorityMetadata(state)
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/noderesources/least_allocated.go]: ditto, this is redundant. please pass a nil and add a comment.

File: pkg/scheduler/framework/plugins/noderesources/most_allocated.go [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+	return MostAllocatedName
+}
+
+// Score invoked at the Score extension point.
+func (ma *MostAllocated) Score(ctx context.Context, state *framework.CycleState, pod *v1.Pod, nodeName string) (int64, *framework.Status) {
+	nodeInfo, exist := ma.handle.NodeInfoSnapshot().NodeInfoMap[nodeName]
+	if !exist {
+		return 0, framework.NewStatus(framework.Error, fmt.Sprintf("node %q does not exist in NodeInfoSnapshot", nodeName))
+	}
+	meta := migration.PriorityMetadata(state)
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/noderesources/most_allocated.go]: ditto, this is redundant, please remove, pass a nil and add a comment.

**Review by ahg-g — COMMENTED** [review]
sorry I missed those.

File: pkg/scheduler/framework/plugins/noderesources/balance_allocated.go (hunk 15 above)
Comment (zouyee) [reply pkg/scheduler/framework/plugins/noderesources/balance_allocated.go]: Done.

**Comment by zouyee** [comment]
/retest
flake https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/84074/pull-kubernetes-kubemark-e2e-gce-big/1185567683546976256/

**Pull request merged**